File: shared/json_recovery.py
"""
Извлечение и разбор структурированных данных из ответов больших языковых моделей.

Общий компонент модуля извлечения характеристик (сервис предобработки, шлюз API, оркестратор):
реализует этапы «выделение JSON-подобного фрагмента» и «парсинг с восстановлением синтаксиса»
многоэтапная процедура: маркер «Ответ:», блок ```json, нормализация кавычек и None→null,
вставка пропущенных запятых, завершение обрезанных структур, запасной разбор при ошибке json.loads.
После разбора признаки проходят семантическую валидацию по схеме справочника в Pydantic уже
в движке правил или сервисе предобработки.
"""

from __future__ import annotations

import ast
import json
import re
from typing import Any


def _extract_json_like(s: str) -> str:
    """
    Вырезает из произвольной строки фрагмент, похожий на JSON.

    Порядок: блок ```json ... ```, иначе текст от первой «{» или «[» до конца строки.
    Не проверяет корректность — только границы кандидата.
    """
    if not isinstance(s, str):
        return ""

    # Сначала ищем оформленный блок кода (частый формат ответов моделей).
    m = re.search(r"```(?:json)?\s*(.*?)\s*```", s, flags=re.IGNORECASE | re.DOTALL)
    if m:
        return m.group(1).strip()

    # Убираем маркеры списков markdown в начале и берём от первой структурной скобки.
    s_stripped = re.sub(r"^[\s\*\-#>]+", "", s.lstrip())
    idx = s_stripped.find("{")
    if idx != -1:
        return s_stripped[idx:].strip()
    idx = s_stripped.find("[")
    if idx != -1:
        return s_stripped[idx:].strip()
    return ""


def _autofix_commas(s: str) -> str:
    """Исправляет типичные синтаксические огрехи: слипшиеся объекты и хвостовые запятые."""
    s = re.sub(r"}\s*{", "}, {", s)
    s = re.sub(r"}\s*\n\s*{", "},\n{", s)
    s = re.sub(r"]\s*{", "], {", s)
    s = re.sub(r",\s*}", "}", s)
    s = re.sub(r",\s*\]", "]", s)
    return s


def _balance_and_close(s: str) -> str:
    """
    Дозакрывает незакрытые кавычки, объекты и массивы в обрезанном ответе модели.

    Обходит строку посимвольно, не считая скобки внутри строковых литералов.
    В конце дописывает недостающие «]» и «}» по накопленной глубине вложенности.
    """
    depth_obj = 0
    depth_arr = 0
    in_string = False
    escape = False

    for ch in s:
        # Внутри строки обратный слэш откладывает смену режима кавычек.
        if ch == "\\" and not escape:
            escape = True
            continue
        elif escape:
            if ch == '"':
                escape = False
                continue
            escape = False

        if ch == '"' and not escape:
            in_string = not in_string
            continue

        # Скобки вне строки меняют счётчики вложенности.
        if not in_string:
            if ch == "{":
                depth_obj += 1
            elif ch == "}":
                if depth_obj > 0:
                    depth_obj -= 1
            elif ch == "[":
                depth_arr += 1
            elif ch == "]":
                if depth_arr > 0:
                    depth_arr -= 1

    # Обрыв посреди строкового значения: закрываем кавычку и при необходимости дописываем значение.
    if in_string:
        s = s + '"'
        if re.search(r'[,{]\s*"[^"]*"$', s.rstrip()):
            s = s.rstrip() + ": null"

    s_stripped = s.rstrip()
    if s_stripped:
        last_non_ws = s_stripped[-1]
        # Обрыв на двоеточии или запятой — подставляем null или убираем висячую запятую.
        if last_non_ws == ":":
            s = s.rstrip() + " null"
        elif last_non_ws == ",":
            s = s_stripped[:-1].rstrip()

    closing = ""
    if depth_arr > 0:
        closing += "]" * depth_arr
    if depth_obj > 0:
        closing += "}" * depth_obj
    if closing:
        s = s + closing
    return s


def parse_json_safe(s: str) -> Any:
    """
    Разбирает строку в dict или list с поэтапной нормализацией.

    Возвращает пустой dict, если вход пустой, фрагмент не найден или разбор не удался.
    Скалярные значения (число, строка без обёртки) не возвращаются — только объект или массив.
    """
    if not isinstance(s, str) or not s.strip():
        return {}

    fragment = _extract_json_like(s)
    if not fragment:
        return {}

    s_clean = fragment
    try:
        # Иногда модель отдаёт буквальные «\\n» вместо перевода строки.
        s_clean = s_clean.replace("\\n", "\n").replace("\\t", "\t").replace("\\r", "\r")
    except Exception:
        pass
    s_clean = s_clean.replace("\r", "").strip()

    # Типографские кавычки и Python None → форма, близкая к JSON.
    s_clean = (
        s_clean.replace("\u201c", '"')
        .replace("\u201d", '"')
        .replace("\u201a", "'")
        .replace("\u2018", "'")
    )
    s_clean = re.sub(r"\bNone\b", "null", s_clean)
    s_clean = _autofix_commas(s_clean)
    s_clean = _balance_and_close(s_clean)
    s_clean = re.sub(r",\s*,+", ",", s_clean)
    s_clean = re.sub(r",\s*}", "}", s_clean)
    s_clean = re.sub(r",\s*\]", "]", s_clean)

    try:
        parsed = json.loads(s_clean)
        if isinstance(parsed, (dict, list)):
            return parsed
        return {}
    except json.JSONDecodeError:
        pass

    # Запасной путь: одинарные кавычки и Python-синтаксис, если JSON строго не прошёл.
    s_eval = re.sub(r"\bnull\b", "None", s_clean)
    try:
        data = ast.literal_eval(s_eval)
        if isinstance(data, (dict, list)):
            return data
        return {}
    except Exception:
        pass
    return {}


def extract_json_from_response(response_text: str) -> str:
    """
    Выделяет JSON-часть из полного текста ответа модели.

    Приоритет:
    1) текст после последнего «Ответ:» / «answer:» (модель часто пишет пояснение, затем JSON);
    2) последний блок ```json ... ``` во всём ответе;
    3) подстрока от первой «{» или «[».
    """
    if not response_text:
        return ""

    response_lower = response_text.lower()
    answer_markers = ("ответ:", "answer:")
    for marker in answer_markers:
        # Берём последнее вхождение маркера — там обычно финальный структурированный ответ.
        last_idx = -1
        search_pos = 0
        while True:
            idx = response_lower.find(marker, search_pos)
            if idx == -1:
                break
            last_idx = idx
            search_pos = idx + 1

        if last_idx != -1:
            json_part = response_text[last_idx + len(marker) :].strip()
            json_part = json_part.lstrip("\n\r\t ")
            json_blocks = list(
                re.finditer(r"```(?:json)?\s*(.*?)\s*```", json_part, flags=re.IGNORECASE | re.DOTALL)
            )
            if json_blocks:
                extracted = json_blocks[-1].group(1).strip()
                if extracted:
                    return extracted
            first_brace = json_part.find("{")
            first_bracket = json_part.find("[")
            start = -1
            if first_brace != -1 and first_bracket != -1:
                start = min(first_brace, first_bracket)
            elif first_brace != -1:
                start = first_brace
            elif first_bracket != -1:
                start = first_bracket
            if start != -1:
                extracted = json_part[start:].strip()
                if extracted:
                    return extracted
            if json_part.strip():
                return json_part

    # Маркера «Ответ:» нет — ищем последний json-блок по всему тексту.
    json_blocks = list(
        re.finditer(r"```(?:json)?\s*(.*?)\s*```", response_text, flags=re.IGNORECASE | re.DOTALL)
    )
    if json_blocks:
        extracted = json_blocks[-1].group(1).strip()
        if extracted:
            return extracted

    first_brace = response_text.find("{")
    first_bracket = response_text.find("[")
    start = -1
    if first_brace != -1 and first_bracket != -1:
        start = min(first_brace, first_bracket)
    elif first_brace != -1:
        start = first_brace
    elif first_bracket != -1:
        start = first_bracket
    if start != -1:
        extracted = response_text[start:].strip()
        if extracted:
            return extracted
    return response_text.strip()


def parse_json_from_model_response(response_text: str) -> Any:
    """
    Точка входа для пайплайна: выделить фрагмент из ответа модели и разобрать его.

    Возвращает dict, list или {} при полной неудаче разбора.
    """
    fragment = extract_json_from_response(response_text)
    return parse_json_safe(fragment)


def is_valid_json_object(s: str) -> bool:
    """Проверяет, что после устойчивого разбора получился непустой объект (словарь)."""
    try:
        parsed = parse_json_safe(s)
        return isinstance(parsed, dict) and bool(parsed)
    except Exception:
        return False

File: frontend/src/utils/jsonRecovery.ts
/**
 * Устойчивое извлечение и парсинг JSON из ячеек датасета / ответов LLM.
 * Согласовано с shared/json_recovery.py (extract_json_from_response + parse_json_safe).
 * JSON5 — гибкий разбор (одинарные кавычки, ключи), аналог fallback ast.literal_eval в Python.
 */
import JSON5 from "json5";

function extractJsonLike(s: string): string {
  if (typeof s !== "string") return "";
  const m = s.match(/```(?:json)?\s*([\s\S]*?)\s*```/i);
  if (m) return m[1].trim();

  const sStripped = s.replace(/^[\s*\-#>]+/, "").trimStart();
  const iObj = sStripped.indexOf("{");
  if (iObj !== -1) return sStripped.slice(iObj).trim();
  const iArr = sStripped.indexOf("[");
  if (iArr !== -1) return sStripped.slice(iArr).trim();
  return "";
}

function autofixCommas(s: string): string {
  let out = s.replace(/}\s*{/g, "}, {");
  out = out.replace(/}\s*\n\s*{/g, "},\n{");
  out = out.replace(/]\s*{/g, "], {");
  out = out.replace(/,\s*}/g, "}");
  out = out.replace(/,\s*]/g, "]");
  return out;
}

function balanceAndClose(s: string): string {
  let depthObj = 0;
  let depthArr = 0;
  let inString = false;
  let escape = false;

  for (let i = 0; i < s.length; i++) {
    const ch = s[i];
    if (ch === "\\" && !escape) {
      escape = true;
      continue;
    }
    if (escape) {
      if (ch === '"') escape = false;
      else escape = false;
      continue;
    }
    if (ch === '"') {
      inString = !inString;
      continue;
    }
    if (!inString) {
      if (ch === "{") depthObj += 1;
      else if (ch === "}") {
        if (depthObj > 0) depthObj -= 1;
      } else if (ch === "[") depthArr += 1;
      else if (ch === "]") {
        if (depthArr > 0) depthArr -= 1;
      }
    }
  }

  let result = s;
  if (inString) {
    result += '"';
    if (/[,{]\s*"[^"]*"$/.test(result.trimEnd())) {
      result = result.trimEnd() + ": null";
    }
  }
  const sStripped = result.trimEnd();
  if (sStripped) {
    const last = sStripped[sStripped.length - 1];
    if (last === ":") result = result.trimEnd() + " null";
    else if (last === ",") result = sStripped.slice(0, -1).trimEnd();
  }

  let closing = "";
  if (depthArr > 0) closing += "]".repeat(depthArr);
  if (depthObj > 0) closing += "}".repeat(depthObj);
  if (closing) result += closing;
  return result;
}

export function extractJsonFromResponse(responseText: string): string {
  if (!responseText) return "";

  const responseLower = responseText.toLowerCase();
  const answerMarkers = ["ответ:", "answer:"] as const;
  for (const marker of answerMarkers) {
    let lastIdx = -1;
    let searchPos = 0;
    for (;;) {
      const idx = responseLower.indexOf(marker, searchPos);
      if (idx === -1) break;
      lastIdx = idx;
      searchPos = idx + 1;
    }
    if (lastIdx !== -1) {
      let jsonPart = responseText.slice(lastIdx + marker.length).trim();
      jsonPart = jsonPart.replace(/^[\n\r\t ]+/, "");
      const jsonBlocks = [...jsonPart.matchAll(/```(?:json)?\s*([\s\S]*?)\s*```/gi)];
      if (jsonBlocks.length) {
        const extracted = jsonBlocks[jsonBlocks.length - 1][1].trim();
        if (extracted) return extracted;
      }
      const firstBrace = jsonPart.indexOf("{");
      const firstBracket = jsonPart.indexOf("[");
      let start = -1;
      if (firstBrace !== -1 && firstBracket !== -1) start = Math.min(firstBrace, firstBracket);
      else if (firstBrace !== -1) start = firstBrace;
      else if (firstBracket !== -1) start = firstBracket;
      if (start !== -1) {
        const extracted = jsonPart.slice(start).trim();
        if (extracted) return extracted;
      }
      if (jsonPart.trim()) return jsonPart;
    }
  }

  const blocks = [...responseText.matchAll(/```(?:json)?\s*([\s\S]*?)\s*```/gi)];
  if (blocks.length) {
    const extracted = blocks[blocks.length - 1][1].trim();
    if (extracted) return extracted;
  }

  const firstBrace = responseText.indexOf("{");
  const firstBracket = responseText.indexOf("[");
  let start = -1;
  if (firstBrace !== -1 && firstBracket !== -1) start = Math.min(firstBrace, firstBracket);
  else if (firstBrace !== -1) start = firstBrace;
  else if (firstBracket !== -1) start = firstBracket;
  if (start !== -1) {
    const extracted = responseText.slice(start).trim();
    if (extracted) return extracted;
  }
  return responseText.trim();
}

/**
 * Как parse_json_safe в Python: сначала extract_json_from_response, затем _extract_json_like, затем нормализация.
 */
export function prepareCleanedJsonString(raw: string): {
  afterResponse: string;
  innerFragment: string;
  sClean: string;
} {
  const afterResponse = extractJsonFromResponse(raw) || raw;
  let innerFragment = extractJsonLike(afterResponse);
  if (!innerFragment) innerFragment = afterResponse.trim();

  let sClean = innerFragment;
  try {
    sClean = sClean.replace(/\\n/g, "\n").replace(/\\t/g, "\t").replace(/\\r/g, "\r");
  } catch {
    /* ignore */
  }
  sClean = sClean.replace(/\r/g, "").trim();

  sClean = sClean
    .replace(/\u201c/g, '"')
    .replace(/\u201d/g, '"')
    .replace(/\u201a/g, "'")
    .replace(/\u2018/g, "'");
  sClean = sClean.replace(/\bNone\b/g, "null");
  sClean = autofixCommas(sClean);
  sClean = balanceAndClose(sClean);
  sClean = sClean.replace(/,\s*,+/g, ",");
  sClean = sClean.replace(/,\s*}/g, "}");
  sClean = sClean.replace(/,\s*]/g, "]");

  return { afterResponse, innerFragment, sClean };
}

function tryParseNormalized(sClean: string): { ok: true; value: unknown } | { ok: false; jsonError: string; json5Error: string } {
  let jsonError = "";
  try {
    return { ok: true, value: JSON.parse(sClean) };
  } catch (e: unknown) {
    jsonError = e instanceof Error ? e.message : String(e);
  }
  const forJson5 = sClean
    .replace(/\bTrue\b/g, "true")
    .replace(/\bFalse\b/g, "false")
    .replace(/\bNone\b/g, "null");
  try {
    return { ok: true, value: JSON5.parse(forJson5) };
  } catch (e: unknown) {
    const json5Error = e instanceof Error ? e.message : String(e);
    return { ok: false, jsonError, json5Error };
  }
}

/**
 * Устойчивый парсинг: как parse_json_safe в Python; при полной неудаче — null (не путать с пустым объектом).
 */
export function parseJsonSafe(s: string): unknown | null {
  if (typeof s !== "string" || !s.trim()) return null;
  const { sClean } = prepareCleanedJsonString(s);
  const r = tryParseNormalized(sClean);
  if (r.ok) return r.value;
  return null;
}

/** Как parse_json_from_model_response в Python: extract + parse_json_safe (внутри prepareCleanedJsonString). */
export function parseJsonFromModelResponse(responseText: string): unknown | null {
  return parseJsonSafe(responseText);
}

export type DatasetParseFailure = {
  ok: false;
  error: string;
  /** Сообщения JSON.parse и JSON5 после нормализации (извлечение фрагмента, кавычки, скобки). */
  parseReason: string;
  /** Ошибка строгого JSON.parse по исходной ячейке (без восстановления). */
  strictCellError: string | null;
  /** Текст после extractJsonFromResponse (как в пайплайне LLM). */
  afterExtractPreview: string;
  /** После extractJsonLike: markdown / первая «{» или «[». */
  innerFragmentPreview: string;
  /** Строка, переданная в JSON.parse / JSON5 после нормализации (усечённо). */
  normalizedAttemptPreview: string;
};

export type DatasetParseResult = { ok: true; value: unknown } | DatasetParseFailure;

const PREVIEW_LEN = 1800;

function strictJsonError(text: string): string | null {
  const t = text.trim();
  if (!t) return null;
  try {
    JSON.parse(t);
    return null;
  } catch (e: unknown) {
    return e instanceof Error ? e.message : String(e);
  }
}

/**
 * Парсинг ячейки JSON датасета с диагностикой при ошибке (как в пайплайне LLM + пояснения для пользователя).
 */
export function parseDatasetFeaturesCell(raw: string): DatasetParseResult {
  const t = raw.trim();
  if (!t) {
    return {
      ok: false,
      error: "пустая ячейка",
      parseReason: "Ячейка пустая или содержит только пробелы.",
      strictCellError: null,
      afterExtractPreview: "",
      innerFragmentPreview: "",
      normalizedAttemptPreview: "",
    };
  }

  const strictCellError = strictJsonError(t);
  const { afterResponse, innerFragment, sClean } = prepareCleanedJsonString(t);
  const attempt = tryParseNormalized(sClean);

  if (attempt.ok) {
    return { ok: true, value: attempt.value };
  }

  const { jsonError, json5Error } = attempt;
  const parseReason = [
    `JSON.parse (после извлечения фрагмента и нормализации): ${jsonError}`,
    `JSON5 (расширенный синтаксис): ${json5Error}`,
  ].join("\n");

  const clip = (s: string) => (s.length > PREVIEW_LEN ? `${s.slice(0, PREVIEW_LEN)}…` : s);

  return {
    ok: false,
    error: "невалидный JSON",
    parseReason,
    strictCellError,
    afterExtractPreview: clip(afterResponse),
    innerFragmentPreview: clip(innerFragment),
    normalizedAttemptPreview: clip(sClean),
  };
}
